rx-java/src/test/java/io/vertx/rx/java/test/WorkerExecutorSchedulerTest.java

```diff
@@ -0,0 +1,254 @@
+  }
+
+  @Test
+  public void testScheduleObserveOnReturnsOnTheCorrectThread() {
+    Observable<String> observable = Observable.<String>create(subscriber -> {
+      assertFalse(Context.isOnVertxThread());
+      subscriber.onNext("expected");
+      subscriber.onCompleted();
+    }).observeOn(scheduler).doOnNext(o -> assertWorkerExecutorThread());
+    new Thread(() -> {
+      observable.subscribe(
+        item -> assertEquals("expected", item),
+        this::fail,
+        this::testComplete);
+    }).start();
+    await();
+  }
+
+  @Test
+  public void testScheduleWithDelayObserveOnReturnsOnTheCorrectThread() {
+    Observable<String> observable = Observable.<String>create(subscriber -> {
+      assertFalse(Context.isOnVertxThread());
+      subscriber.onNext("expected");
+      subscriber.onCompleted();
+    }).delay(10, TimeUnit.MILLISECONDS, scheduler).doOnNext(o -> assertWorkerExecutorThread());
+    new Thread(() -> {
+      observable.subscribe(
+        item -> assertEquals("expected", item),
+        this::fail,
+        this::testComplete);
+    }).start();
+    await();
+  }
+
+  @Test
+  public void testScheduleDelayed() {
+    Scheduler.Worker worker = scheduler.createWorker();
+    long time = System.currentTimeMillis();
+    worker.schedule(() -> {
+      assertWorkerExecutorThread();
+      assertTrue(System.currentTimeMillis() - time >= 40);
+      testComplete();
+    }, 40, TimeUnit.MILLISECONDS);
+    await();
+  }
+
+  @Test
+  public void testSchedulePeriodic() {
+    Scheduler.Worker worker = scheduler.createWorker();
+    AtomicLong time = new AtomicLong(System.currentTimeMillis() - 40);
+    AtomicInteger count = new AtomicInteger();
+    AtomicReference<Subscription> sub = new AtomicReference<>();
+    sub.set(worker.schedulePeriodically(() -> {
+      assertWorkerExecutorThread();
+      if (count.incrementAndGet() > 2) {
+        sub.get().unsubscribe();
+        testComplete();
+      } else {
+        long now = System.currentTimeMillis();
+        long delta = now - time.get();
+        assertTrue("" + delta, delta >= 40);
+        time.set(now);
+      }
+    }, 0, 40, TimeUnit.MILLISECONDS));
+    await();
+  }
+
+  @Test
+  public void testUnsubscribeBeforeExecute() throws Exception {
+    Scheduler.Worker worker = scheduler.createWorker();
+    CountDownLatch latch = new CountDownLatch(1);
+    Subscription sub = worker.schedule(latch::countDown, 20, TimeUnit.MILLISECONDS);
+    sub.unsubscribe();
+    assertFalse(latch.await(40, TimeUnit.MILLISECONDS));
+  }
+
+  @Test
+  public void testUnsubscribeDuringExecute() throws Exception {
+    Scheduler.Worker worker = scheduler.createWorker();
+    AtomicInteger count = new AtomicInteger();
+    AtomicReference<Subscription> sub = new AtomicReference<>();
+    sub.set(worker.schedulePeriodically(() -> {
+      if (count.getAndIncrement() == 0) {
+        sub.get().unsubscribe();
+      }
+    }, 0, 5, TimeUnit.MILLISECONDS));
+    Thread.sleep(60);
```

Comment:
Latch should be used in favor of sleep when trying to synchronize between
different routines, yes. But here there is no event we're waiting for (see
my other reply)

2016-08-11 1:38 GMT+02:00 testn <notifications@github.com>:

> In rx-java/src/test/java/io/vertx/rx/java/test/
> WorkerExecutorSchedulerTest.java
> <https://github.com/vert-x3/vertx-rx/pull/46#discussion_r74350986>:
>
> > +    Subscription sub = worker.schedule(latch::countDown, 20, TimeUnit.MILLISECONDS);
> > +    sub.unsubscribe();
> > +    assertFalse(latch.await(40, TimeUnit.MILLISECONDS));
> > +  }
> > +
> > +  @Test
> > +  public void testUnsubscribeDuringExecute() throws Exception {
> > +    Scheduler.Worker worker = scheduler.createWorker();
> > +    AtomicInteger count = new AtomicInteger();
> > +    AtomicReference<Subscription> sub = new AtomicReference<>();
> > +    sub.set(worker.schedulePeriodically(() -> {
> > +      if (count.getAndIncrement() == 0) {
> > +        sub.get().unsubscribe();
> > +      }
> > +    }, 0, 5, TimeUnit.MILLISECONDS));
> > +    Thread.sleep(60);
>
> latch should be used in favor of sleep
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-rx/pull/46/files/e7488c91f845d8c98eaad4b9bb16839546e7b18a#r74350986>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABblttPjjEmfUlngVlkJrd2zgJ4SUxv1ks5qemDggaJpZM4JhJfH>
> .
>